Assets/Weapons/Weapon.cs

```diff
@@ -0,0 +1,135 @@
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+/// <summary>
+/// Classify different weapon types
+/// </summary>
+public enum EWeaponType
+{
+    Fire,
+    Melee,
+    Default
+}
+
+/// <summary>
+/// This interface will implement all the functions weapons will make use of
+/// </summary>
+public interface IWeapon
+{
+    public EWeaponType GetWeaponType();
+
+    /// <summary>
+    /// Returns the amount of uses (bullets left), -1 if this weapon has no Uses functionality 
+    /// </summary>
+    public int UsesLeft()
+    {
+        return -1;
+    }
+
+    public void Pickup(Transform weaponHolder);
+
+    public void Throw(Vector2 forwardVector);
+
+    public void Use(bool pressed);
+}
+
+/// <summary>
+/// Base weapon class, basic functionality
+/// </summary>
+public class Weapon : MonoBehaviour, IWeapon
+{
+    [Header("Base")] 
+    public EWeaponType weaponType = EWeaponType.Default;
+    
+    public float throwForce = 15;
+    
+    public float rotationForce = 1000;
+
+    public Collider2D gfxCollider2D;
+
+    private bool _held;
+    private Rigidbody2D _rb;
+    
+    public virtual int UsesLeft()
+    {
+        return -1;
+    }
+
+    /// <returns>Returns this weapon type</returns>
+    public virtual EWeaponType GetWeaponType()
+    {
+        return weaponType;
+    }
+
+    public virtual void Start()
```

Comment:
Make functions protected for all the things that reeas to be virtual but thatdon't need to be public

```suggestion
    protected virtual void Start()
```